packages/store/src/workspace/page.ts

```diff
@@ -389,44 +389,65 @@ export class Page extends Space<FlatBlockMap> {
       return;
     }
 
-    const firstBlock = blocks[0];
-    const currentParent = this.getParent(firstBlock);
-
-    // the blocks must have the same parent (siblings)
-    if (blocks.some(block => this.getParent(block) !== currentParent)) {
-      console.error('the blocks must have the same parent');
-    }
-
-    if (currentParent === null || newParent === null) {
+    if (newParent === null) {
       throw new Error("Can't find parent model");
     }
 
+    const tempMap = new Map<BaseBlockModel, BaseBlockModel[]>();
+
     blocks.forEach(block => {
+      const parent = this.getParent(block);
+
+      if (parent === null) {
+        throw new Error("Can't find parent model");
+      }
+
       this.schema.validate(block.flavour, newParent.flavour);
+
+      const subArray = tempMap.get(parent);
+      if (subArray) {
+        subArray.push(block);
+      } else {
+        tempMap.set(parent, [block]);
+      }
     });
 
     this.transact(() => {
-      const yParentA = this._yBlocks.get(currentParent.id) as YBlock;
-      const yChildrenA = yParentA.get('sys:children') as Y.Array<string>;
-      const idx = yChildrenA.toArray().findIndex(id => id === firstBlock.id);
-      yChildrenA.delete(idx, blocks.length);
       const yParentB = this._yBlocks.get(newParent.id) as YBlock;
       const yChildrenB = yParentB.get('sys:children') as Y.Array<string>;
 
-      let nextIdx = 0;
-      if (newSibling) {
-        nextIdx = yChildrenB.toArray().findIndex(id => id === newSibling.id);
-      }
+      // after the target
+      let offset = 1;
 
-      const ids = blocks.map(block => block.id);
-      if (insertBeforeSibling) {
-        yChildrenB.insert(nextIdx, ids);
-      } else {
-        yChildrenB.insert(nextIdx + 1, ids);
+      for (const [parent, blocks] of tempMap) {
```

Comment:
It seems that each array value of the `tempMap` here only has one item (line 411), when do we need to loop through the `blocks` here?

Reply:
This is the case in the `heavy-whiteboard` example, but we can edit the frames.